### PR TITLE
cloud-init schema validation

### PR DESCRIPTION
**Pull request by monstermunchkin**
This validates cloud-init config keys, and ensures they are valid cloud-config or YAML.

### CODE REVIEW

**Review by tomponline — APPROVED** [review]

File: shared/validate/validate.go [hunk 1]
```diff
@@ -787,3 +788,39 @@ func IsRequestURL(value string) error {
 
 	return nil
 }
+
+// IsCloudConfig checks value is a valid cloud-config.
+func IsCloudConfig(value string) error {
+	if value == "" {
```
Comment (tomponline) [inline shared/validate/validate.go:794]: This part shouldn't be needed as your using validate.Optional

File: shared/validate/validate.go [hunk 2]
```diff
@@ -787,3 +788,39 @@ func IsRequestURL(value string) error {
...
+	if len(lines) == 1 {
+		return nil
+	}
+
+	return IsYAML(lines[1])
+}
+
+// IsYAML checks value is valid YAML.
+func IsYAML(value string) error {
+	if value == "" {
```
Comment (tomponline) [inline shared/validate/validate.go:814]: This part shouldn't be needed as your using validate.Optional

**Review by tomponline — CHANGES_REQUESTED** [review]
Just a nit

**Review by tomponline — APPROVED** [review]

**Comment by tomponline** [comment]
Tests not happy

File: shared/validate/validate.go [hunk 3]
```diff
@@ -787,3 +788,31 @@ func IsRequestURL(value string) error {
...
 	return nil
 }
+
+// IsCloudConfig checks value is a valid cloud-config.
+func IsCloudConfig(value string) error {
+	lines := strings.SplitN(value, "\n", 2)
+
+	if lines[0] != "#cloud-config" {
+		return fmt.Errorf("Missing cloud-config header")
+	}
```
Comment (stgraber) [inline shared/validate/validate.go:798]: We shouldn't do that. There are plenty of other valid formats for cloud-init other than `#cloud-config`.

Anything other than `#cloud-config` should be considered valid and be left alone.

Comment (monstermunchkin) [reply shared/validate/validate.go]: I've updated the PR to only validate YAML if the `#cloud-config` header is present.

**Pull request merged**